File: lib/camus-sync/slot_options.rb
require 'fileutils'
require 'webhdfs'
require 'webhdfs/fileutils'
require 'camus-sync/slot'
require 'camus-sync/time_ext'

module CamusSync
  class SlotOptions
    attr_reader :topic, :time

    def initialize(camus_path, hdfs_servers, topic, time)
      @camus_path = camus_path
      @hdfs_servers = hdfs_servers
      @topic = topic
      @time = time
      @all = all!
    end

    def all!
      @hdfs_servers.map do |hdfs|
        Slot.new(@camus_path, hdfs, @topic, @time)
      end
    end

    def best_slot
      @all.max_by(&:events)
    end

    def synchronize!(dryrun = true)
      $log.info("synchronizing slot options at", time: @time, topic: @topic)

      unless @hdfs_servers.size > 1
        $log.info("tried to sync but there are no enough hdfs servers")
        return
      end

      all_events = @all.map(&:events).uniq
      unless all_events.size > 1
        $log.info("slot already in sync", time: @time, source: @topic)
        return
      end

      unless dryrun
        $log.info("downloading best option", time: @time, from: best_slot.hdfs.host)
        WebHDFS::FileUtils.set_server(best_slot.hdfs.host, 50070) unless dryrun

        FileUtils.remove_entry_secure LOCAL_TMP_DIR, true
        FileUtils.mkdir LOCAL_TMP_DIR

        best_slot.download
      end

      @all.each do |option|
        next if option == best_slot || option.events == best_slot.events
        $log.info("found differences between options", delta: (best_slot.events - option.events), best: best_slot.events, current: option.events)
        next if dryrun
        
        $log.info("|-- synchronizing", from: best_slot.hdfs.host, to: option.hdfs.host)
        option.destroy
        option.upload best_slot.paths.map { |p| p.split('/')[-1] }
        $log.info("|-- sync done", at: option.hdfs.host)
      end
    end 
  end
end

File: src/main/java/net/redborder/camus/SlotOptions.java
package net.redborder.camus;

import org.apache.hadoop.fs.Path;
import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

public class SlotOptions {
    private static final Logger log = LoggerFactory.getLogger(SlotOptions.class);

    private final String camusPath;
    private final List<HdfsServer> servers;
    private final List<Slot> slots;
    private final Slot bestSlot;
    private final String topic;
    private final DateTime time;

    public SlotOptions(String camusPathStr, List<HdfsServer> servers, String topic, DateTime time) {
        this.camusPath = camusPathStr;
        this.servers = servers;
        this.topic = topic;
        this.time = time;
        this.slots = loadSlots();
        this.bestSlot = Collections.max(slots);
    }

    public String getTopic() {
        return topic;
    }

    public DateTime getTime() {
        return time;
    }

    private List<Slot> loadSlots() {
        List<Slot> createdSlots = new ArrayList<>();

        for (HdfsServer server: servers) {
            createdSlots.add(new Slot(camusPath, server, topic, time));
        }

        return createdSlots;
    }

    public void deduplicate(boolean dryrun, List<String> dimensions) {
        log.info("Deduplicate slot for topic {} at time {}", topic, time);
        List<String> paths = new ArrayList<>();
        Set<Long> randomNumberSet = new HashSet<>();
        Set<Long> eventsSet = new HashSet<>();

        for (Slot slot : slots) {
            if (slot.getPaths().isEmpty()) {
                log.info("|-- [!!] Empty or missing slot at path {}", slot.getFullFolder());
            } else {
                log.info("|-- Found slot at path {}", slot.getFullFolder());
                paths.add(slot.getFullFolder());
            }

            randomNumberSet.add(slot.getRandomNumber());
            eventsSet.add(slot.getEvents());
        }

        if (paths.isEmpty()) {
            log.info("No paths available to deduplicate, ignoring...");
        } else if ((randomNumberSet.size() == 1 && eventsSet.size() == 1)) {
            log.info("Slot for topic {} time {} is already deduplicated, ignoring...", topic, time);
        } else if (!dryrun) {
            DeduplicationJob pigJob = new DeduplicationJob(paths, dimensions);
            DeduplicationJob.Results results = pigJob.run();
            log.info("Written {} records into {}", results.getNumberRecords(), results.getPath().toString());

            if (results.getNumberRecords() == 0) {
                log.error("Pig deduplicate job did not write a damn single thing!");
                log.error("Aborting deduplicate job for topic {} time {}", topic, time);
                return;
            }

            UUID uuid = UUID.randomUUID();
            long randomNumber = Math.abs(uuid.getMostSignificantBits());

            for (Slot slot : slots) {
                slot.destroy();
                String uploadName = topic + ".0.0." + String.valueOf(results.getNumberRecords()) + "." + randomNumber + ".deduplicated.gz";
                slot.upload(results.getPath(), uploadName);
            }
        }
    }

    public void synchronize(boolean dryrun) {
        log.info("Synchronizing slot for topic {} at time {}", topic, time);

        Set<Long> events = new HashSet<>();
        for (Slot slot : slots) {
            events.add(slot.getEvents());
        }

        if (events.size() < 2) {
            log.info("Slot already in sync topic {} time {}", topic, time);
            return;
        }

        for (Slot slot : slots) {
            if (bestSlot == slot || bestSlot.getEvents() == slot.getEvents()) continue;
            log.info("Found differences between options | delta {} best {} current {}",
                    (bestSlot.getEvents() - slot.getEvents()),
                    bestSlot.getEvents(),
                    slot.getEvents());

            if (dryrun) continue;

            log.info("|-- Synchronizing from {} to {}", bestSlot.getServer().getHostname(), slot.getServer().getHostname());
            slot.destroy();

            for (Path path : bestSlot.getPaths()) {
                slot.upload(path);
            }

            log.info("|-- Sync done at {}", slot.getServer().getHostname());
        }
    }
}
